From d6f761febd18e5e9f408ffe7343f2e39e1047863 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Wed, 7 Jan 2015 16:04:57 -0500
Subject: [iommu] vt-d: Fix dmar_domain leak in iommu_attach_device

Message-id: <20150107160457.9952.48018.stgit@amt.stowe>
Patchwork-id: 102520
O-Subject: [RHEL7.1 PATCH 3/3] iommu/vt-d: Fix dmar_domain leak in iommu_attach_device
Bugzilla: 1109829
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1109829
Upstream Status: 62c22167dd70b730f61c2b88f950e98154a87980
>From Joerg's iommu/fixes branch intended for v3.19 -
https://git.kernel.org/cgit/linux/kernel/git/joro/iommu.git/commit/?h=iommu/fixes&id=62c22167dd70b730f61c2b88f950e98154a87980

Conflict:
  This patch has a dependency on upstream commit ab8dfe2 ("iommu/vt-d:
  Introduce helper functions to improve code readability") which was
  introduced during v3.17-rc1.  RHEL7 does not have that commit, and thus
  does not have 'domain_type_is_vm_or_si()'.  As such, this back-port uses
  the older method used prior to the introduction of
  'domain_type_is_vm_or_si() (see upstream commit ab8dfe2).

commit 62c22167dd70b730f61c2b88f950e98154a87980
Author: Joerg Roedel <jroedel@suse.de>
Date:   Tue Dec 9 12:56:45 2014 +0100

    iommu/vt-d: Fix dmar_domain leak in iommu_attach_device

    Since commit 1196c2f a domain is only destroyed in the
    notifier path if it is hot-unplugged. This caused a
    domain leakage in iommu_attach_device when a driver was
    unbound from the device and bound to VFIO. In this case the
    device is attached to a new domain and unlinked from the old
    domain. At this point nothing points to the old domain
    anymore and its memory is leaked.
    Fix this by explicitly freeing the old domain in
    iommu_attach_domain.

Fixes: 1196c2f (iommu/vt-d: Fix dmar_domain leak in iommu_attach_device)
Cc: <stable@vger.kernel.org> # v3.18
Tested-by: Jerry Hoemann <jerry.hoemann@hp.com>
Signed-off-by: Joerg Roedel <jroedel@suse.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/intel-iommu.c b/drivers/iommu/intel-iommu.c
index b2f9aa7..1306ea9 100644
--- a/drivers/iommu/intel-iommu.c
+++ b/drivers/iommu/intel-iommu.c
@@ -4270,6 +4270,11 @@ static int intel_iommu_attach_device(struct iommu_domain *domain,
     domain_remove_one_dev_info(old_domain, dev);
    else
     domain_remove_dev_info(old_domain);
+
+   if (!(old_domain->flags & DOMAIN_FLAG_VIRTUAL_MACHINE) &&
+       !(old_domain->flags & DOMAIN_FLAG_STATIC_IDENTITY) &&
+       list_empty(&old_domain->devices))
+    domain_exit(old_domain);
   }
  }
 
-- 
1.7.1